src/cliffordsynthesis/CliffordSynthesizer.cpp

```diff
@@ -395,4 +416,82 @@ void CliffordSynthesizer::updateResults(const Configuration& config,
     break;
   }
 }
+
+// assume canonical sorting of gates
+std::vector<std::size_t> getLayers(const qc::QuantumComputation& qc) {
+  std::vector<std::size_t> layerNum{};
+  layerNum.resize(qc.size());
+  std::vector<std::size_t> layers{};
+  std::size_t              layer = 0U;
+  for (std::size_t i = 0; i < qc.size(); ++i) {
+    const auto& gate = *(qc.begin() + i);
+
+    for (const auto& qubit : gate->getUsedQubits()) {
+      if (layerNum[qubit] >= layer) {
+        ++layer;
+        layers.emplace_back(i);
+        break;
+      }
+    }
+    for (const auto& qubit : gate->getUsedQubits()) {
+      layerNum[qubit] = layer;
+    }
+  }
+  if (layers.back() < qc.size()) {
+    layers.emplace_back(qc.size());
+  }
+  return layers;
+}
+
+void CliffordSynthesizer::depthHeuristicSynthesis() {
+  INFO() << "Optimizing Circuit with Heuristic" << std::endl;
+  auto optimalConfig                 = configuration;
+  optimalConfig.heuristic            = false;
+  optimalConfig.target               = TargetMetric::Depth;
+  optimalConfig.initialTimestepLimit = configuration.splitSize;
+
+  qc::CircuitOptimizer::reorderOperations(*initialCircuit);
+  qc::QuantumComputation   optCircuit{initialCircuit->getNqubits()};
+  std::vector<std::size_t> layers = getLayers(*initialCircuit);
+
+  std::vector<std::future<std::shared_ptr<qc::QuantumComputation>>> subCircuits;
+  for (std::size_t i = 0; i < layers.size() - 1; i += configuration.splitSize) {
+    std::size_t const          startIdx = layers[i];
+    std::optional<std::size_t> endIdx;
+
+    if (i + configuration.splitSize >= layers.size()) {
+      endIdx = layers.back();
+    } else {
+      endIdx = layers[i + configuration.splitSize];
+    }
+
+    // launch threads
+    subCircuits.emplace_back(std::async(
+        std::launch::async | std::launch::deferred,
+        [this, startIdx, endIdx, &optimalConfig]() {
+          return cs::CliffordSynthesizer::synthesizeSubcircuit(
+              initialCircuit, startIdx, endIdx.value(), optimalConfig);
+        }));
+  }
+
+  for (auto& subCircuit : subCircuits) {
+    const auto& circ = subCircuit.get();
+    for (const auto& op : *circ) {
+      optCircuit.emplace_back(op->clone());
+    }
+  }
```

Comment:
This seems a bit wasteful to me as it incurs real copies of all operations. The sub circuits are not needed any more, so their operations can simply be moved to the optimal circuit without incurring copies.